Add various features to aio-s3

thehesiod opened the pull request:
- fix for buckets with "." in their name
- add multipart upload from existing s3 file
- add support for getting “head”
- add https support
- add support for bucket location
- add list of existing multipart uploads
- use botocore auth + signing to pull jam rules/environment/etc.
-  switch to xmltodict for parsing
- add initial support for logging
- switch to use connection pooling to reduce overhead
- fix multipart bugs to support adding parts out of order and multiple parts in parallel
- added retry support
- centralized request code to yield for data, do retries and handle error codes

note: a few of these are not backwards compatible

Commit:
Adding support for doing a multipart upload from an existing S3 object.

Commit:
Adding exception specific to 404 status code.

Commit:
Add support for using https to talk to S3.

Commit:
Pool http connections per bucket.

Commit:
Waiting for close to complete

Commit:
Add context to exception messages.

Commit:
Adding function to get S3 bucket region.

Commit:
bug fix.

Commit:
Adding support for downloading a specific version of an object.

Commit:
Adding support for copying a specific version of an object.

Commit:
URL encoding source key name.

thehesiod committed:
enhancements

- switch to xmltodict to make debugging/coding simpler
- added support for setting scheme of requests
- added multipart copy_chunk support
- fixed multipart “part” numbering due to scoping and sorting issues
- added credential resolver support
- added head, abort_multipart, list_multipart_uploads, upload_file, and
copy support
- added support to set metadata on upload_multipart

thehesiod committed:
Merge branch 's3s3' into amohr-work

Conflicts:
	aios3/bucket.py

thehesiod committed:
Merge pull request #1 from farmersbusinessnetwork/s3s3

integrate s3s3 to master

thehesiod committed:
Merge branch 'master' into amohr-s3

thehesiod committed:
merge fixes

thehesiod committed:
more merge fixes

thehesiod committed:
more merge clean-up/fixes

thehesiod committed:
one more fix

cbrichford committed:
Merge pull request #2 from farmersbusinessnetwork/amohr-s3

enhancements

thehesiod committed:
more modernization of aio-s3

- removed v2 signature support since v4 is now supported on all APIs
across all regions
- added logging support
- added ability to add multipart chunks out of order
- collapsed Request class in to Bucket._request
- added retry logic as the AWS APIs are really finicky.
- collapsed error checking / data loading into Bucket._request
- added “scheme” support to fix the port ugliness for https, plus https
can be on ports other than 443
- updated keep alive to keep AWS happy
- moved getLocation to Bucket
- simplified header logic
- switched error to also use xmltodict

thehesiod committed:
update version

cbrichford committed:
Merge pull request #3 from farmersbusinessnetwork/amohr-s3

lots more goodies

thehesiod committed:
this is a warning, not an error

thehesiod committed:
Merge pull request #4 from farmersbusinessnetwork/amohr-s3

this is a warning, not an error

thehesiod committed:
fix for buckets with "." in their name

- switches to use botocore’s auth module so we’re not
replicating/maintaining signing methods
- switches to path style API calls to support buckets with names
- gets rid of port parameter since s3 does not listen on ports besides
80/443.

thehesiod committed:
update version

thehesiod committed:
Merge branch 'master' into amohr-s3

thehesiod committed:
add new dependency

thehesiod committed:
update version for new requirement

cbrichford committed:
Merge pull request #5 from farmersbusinessnetwork/amohr-s3

fix for buckets with "." in their name

tailhook commented on aios3/bucket.py:
```diff
@@ -13,312 +13,210 @@
+
+        x = xmltodict.parse(data)['ListBucketResult']
+
+        if 'IsTruncated' != 'false':
```
Looks like a typo

thehesiod replied:
thanks!  submitted fix, and also fixed support of old aws_key/aws_secret.

thehesiod committed:
bugfixes

- fix IsTruncated copy/paste
- fix old aws_key/aws_secret parameter handling

thehesiod committed:
fix requirements

tailhook commented:
Well, it seems I don't have enough time to maintain the project anymore. You may either use your fork or switch to aiobotocore (I've not tried it myself). Sorry.

thehesiod committed:
add timeout support

thehesiod committed:
tweak to timeout

thehesiod committed:
Merge pull request #6 from farmersbusinessnetwork/amohr-timeout

add timeout support

thehesiod committed:
fix CTRL+C

thehesiod committed:
Merge pull request #7 from farmersbusinessnetwork/amohr-timeout

fix CTRL+C

thehesiod committed:
switch to python3.5 to simplify + robustify

thehesiod committed:
Merge pull request #8 from farmersbusinessnetwork/amohr-timeout

switch to python3.5 to simplify + robustify

thehesiod committed:
add delimiter support and switch list_by_chunks to "async for"

thehesiod committed:
Merge pull request #9 from farmersbusinessnetwork/amohr-timeout

add delimiter support and switch list_by_chunks to "async for"

thehesiod committed:
aios3 updates

- add prelim presign support
- add stats class
- cache signer
- return headers for get/put
- ensure CommonPrefixes + Contents always return as a list

thehesiod committed:
Merge pull request #10 from farmersbusinessnetwork/amohr-timeout

aios3 updates

thehesiod committed:
round floats + exponential slowdown

thehesiod committed:
Merge pull request #11 from farmersbusinessnetwork/amohr-timeout

round floats + exponential slowdown

thehesiod committed:
add boto_style to list

thehesiod committed:
Merge pull request #12 from farmersbusinessnetwork/amohr-timeout

add boto_style to list

thehesiod committed:
- make list_by_chunks use list
- decode the list results

thehesiod committed:
using aliases for now

thehesiod committed:
actually retry errors ;)

thehesiod committed:
fix retry logic (404s should go through)

fix bug in list

thehesiod committed:
Merge pull request #13 from farmersbusinessnetwork/amohr-timeout

fixing issues

thehesiod committed:
add timout init and delete_objects support

thehesiod committed:
Merge pull request #14 from farmersbusinessnetwork/amohr-timeout

add timout init and delete_objects support

thehesiod committed:
add uri encode to call

thehesiod committed:
Merge pull request #15 from farmersbusinessnetwork/amohr-timeout

add uri encode to call

thehesiod committed:
updates

- implement default credential resolver
- fix a few list issues

thehesiod committed:
Merge pull request #16 from farmersbusinessnetwork/amohr-timeout

updates

thehesiod committed:
do it in a safer way

thehesiod committed:
Merge pull request #17 from farmersbusinessnetwork/amohr-timeout

do what the docs say

thehesiod committed:
retry on 500 as well

thehesiod committed:
Merge pull request #18 from farmersbusinessnetwork/amohr-timeout

retry on 500 as well

thehesiod committed:
actually retry RuntimeErrors generated by errors.AWSException.from_bytes

thehesiod committed:
Merge pull request #19 from farmersbusinessnetwork/amohr-timeout

actually retry RuntimeErrors generated by errors.AWSException.from_bytes

thehesiod committed:
fix getLocation + paging of list multipart uploads

thehesiod committed:
Merge pull request #20 from farmersbusinessnetwork/amohr-timeout

fix getLocation + paging of list multipart uploads

thehesiod committed:
another region fix

thehesiod committed:
Merge pull request #21 from farmersbusinessnetwork/amohr-timeout

another region fix

thehesiod committed:
always urlencode the url

we need to do this here due so the signature will match the actual
request

thehesiod committed:
Merge pull request #22 from farmersbusinessnetwork/amohr-timeout

always urlencode the url

thehesiod committed:
simplify code

thehesiod committed:
Merge pull request #23 from farmersbusinessnetwork/amohr-timeout

simplify code

thehesiod committed:
don't swallow CancelledError

thehesiod committed:
Merge pull request #24 from farmersbusinessnetwork/amohr-timeout

don't swallow CancelledError

thehesiod committed:
retry ALL s3 errors as we can't really trust any

thehesiod committed:
Merge pull request #25 from farmersbusinessnetwork/amohr-timeout

retry ALL s3 errors as we can't really trust any

thehesiod committed:
switch to botocore response parser

thehesiod committed:
Merge branch 'master' into amohr-timeout

thehesiod committed:
bugfixes

thehesiod committed:
remove old comment

thehesiod committed:
bugfix

thehesiod committed:
updates

- add support for list_object_versions
- make list APIs more like botocore

thehesiod committed:
more botocoreification

thehesiod committed:
still need xml2dict to generate xml

thehesiod committed:
sigh

thehesiod committed:
bugfix, default is already correct

thehesiod committed:
make this method global

thehesiod committed:
Merge pull request #26 from farmersbusinessnetwork/amohr-timeout

add RetryHandler and switch to botocore response parser

thehesiod committed:
bugfix

thehesiod committed:
Merge pull request #27 from farmersbusinessnetwork/amohr-timeout

bugfix

thehesiod committed:
bugfixes

thehesiod committed:
Merge pull request #28 from farmersbusinessnetwork/amohr-timeout

bugfixes

thehesiod committed:
bugfixes for aio-s3

get_bucket_location had a special post-parse parser, and upload_part
had two cases

thehesiod committed:
Merge pull request #29 from farmersbusinessnetwork/amohr-timeout

bugfixes for aio-s3

thehesiod committed:
add a few new methods for weather bucket backup

thehesiod committed:
Merge pull request #30 from farmersbusinessnetwork/amohr-timeout

add a few new methods for weather bucket backup

thehesiod committed:
add support for dict params and fix encoding

thehesiod committed:
Merge pull request #31 from farmersbusinessnetwork/amohr-timeout

add support for dict params and fix encoding

thehesiod committed:
bugfix

thehesiod committed:
Merge pull request #32 from farmersbusinessnetwork/amohr-timeout

bugfix

thehesiod committed:
correctly encode parameters

thehesiod committed:
bump up ver

thehesiod committed:
Merge pull request #33 from farmersbusinessnetwork/amohr-timeout

correctly encode parameters
